virtio: centralize mmio slot handling

luigix25 opened the pull request:
Currently the (only) virtio driver does the slot probing by itself. This is problematic when there are multiple drivers in the system.

Introduce `MMIOSlot`. This structure is in charge of probing all the slots and enumerate them.

As of now, this only supports `virtio-mmio` slots with _one_ device per type.

stefano-garzarella commented on kernel/src/svsm.rs:
```diff
@@ -151,6 +153,27 @@ fn mapping_info_init(launch_info: &KernelLaunchInfo) {
+        use svsm::block::virtio_blk::initialize_block;
+
+        let Some(block_slot) = slots.pop_slot(virtio_drivers::transport::DeviceType::Block) else {
+            log::warn!("MmioSlots: no virtio-blk device found");
```
why a warning here?
Is it always expected to have a virtio-blk device?

luigix25 replied:
> Maybe instead of warning, I'll print info when we find them.

this already happens in `probe_mmio_slots` when a slot is successfully probed. 
I'll switch the log to info from debug.

stefano-garzarella commented on kernel/src/virtio/slot.rs:
```diff
@@ -0,0 +1,112 @@
+/// The caller must ensure:
+/// * Single-threaded execution
+/// * Called exactly once during boot
+pub unsafe fn probe_mmio_slots() -> Option<MmioSlots> {
```
Why an Option? Can be just an empty Vec ?

stefano-garzarella commented on kernel/src/block/virtio_blk.rs:
```diff
@@ -147,11 +142,14 @@ mod tests {
         let expected_hash = get_image_hash_from_host().unwrap();
 
-        let n_sectors = blk.size() / SECTOR_SIZE;
+        let n_sectors = blk_driver.size() / SECTOR_SIZE;
```
I'd avoid this renaming in this commit to keep it as small as possible.

stefano-garzarella commented on kernel/src/virtio/slot.rs:
```diff
@@ -0,0 +1,112 @@
+}
+
+impl MmioSlots {
+    /// Retrieves and removes the first available MMIO slot matching the specified device type.
```
From https://doc.rust-lang.org/std/vec/struct.Vec.html#method.pop_if

> Removes and returns the last element from a vector if the predicate returns true ...

So, are you sure this comment is true?
Maybe a test will be nice to have.

stefano-garzarella commented on kernel/src/block/api.rs:
> Send + Sync traits are required for concurrency handling.

This is not clear to me, mainly IIUC this is required by ImmutAfterInitCell()  no?

stefano-garzarella replied:
Okay, so why not just writing that `ImmutAfterInitCell` requires that?

luigix25 commented on kernel/src/virtio/slot.rs:
```diff
@@ -0,0 +1,121 @@
+///
+/// # Safety
+///
+/// Must be called on the bootstrap processor (BSP) before other CPUs are started.
```
well, it would read again _all_ the slots from the FwCfg interface. So the same memory areas would be mapped again. I'm not sure what would happen if you try to instantiate a second driver on the same virtio-device, but we agree that this must not happen.

stefano-garzarella replied:
Okay, but what Rust memory-safety guarantees we are violating if that happen?
This is what a Safety section should talk about.

luigix25 replied:
I guess none. I'll remove the `unsafe`.
Should I leave the comments? (without the safety keyword)

stefano-garzarella replied:
Really, I don't see any issue on running probe_mmio_slots() concurrently and also twice or more, so I don't understand what you are worried about for both "must":

> /// The caller must ensure:
> /// * Single-threaded execution
> /// * Called exactly once during boot

The main actions done here are:
- `map_global_range_4k_shared()` will map the same phys_addr on multiple virtual address (yeah, it's not great, but what will be the issue?)
- `MmioTransport::<SvsmHal>::new()` does only reads, so again what will be the issue?

I agree that doesn't make sense to call it twice, but for sure nothing wrong can happen, or I may miss something.

stefano-garzarella commented on kernel/src/block/mod.rs:
```diff
@@ -10,3 +10,9 @@ pub mod error;
+use crate::{block::api::BlockDriver, utils::immut_after_init::ImmutAfterInitCell};
+use alloc::boxed::Box;
+
+static BLOCK_DEVICE: ImmutAfterInitCell<Box<dyn BlockDriver>> = ImmutAfterInitCell::uninit();
```
I think we should also document this in Documentation/ somewhere, but we can do later maybe with #806

stefano-garzarella commented on kernel/src/svsm.rs:
```diff
@@ -151,6 +153,22 @@ fn mapping_info_init(launch_info: &KernelLaunchInfo) {
+
+        let Some(block_slot) = slots.pop_slot(virtio_drivers::transport::DeviceType::Block) else {
+            return;
+        };
```
I'd move this into the `initialize_block()`, so if in the future we will handle multiple block devices, we can do everything there, in the block module.

luigix25 commented:
v4:
- addressed @stefano-garzarella comments.
- Better error handling
- Removed unnecessary changes
- Better documentation
- block driver now consumes the slot
